### PR TITLE
feat!: inter auction status command

### LINKED ISSUE #7266: [CLI] Auction Status interface
## Describe the bug
As a bidding user, I want to list the remaining time left on an auction and the current price of the auction, so that I can best inform my bidding strategy.

## To Reproduce
1. Bid on some liquidating vaults
2. Unable to easily see other bids I'm competing with, or how long there is for an auction I have already bid on, or if my bid will currently win the auction at the end of the duration

## Expected behavior
A user can see if their bid is the current winning bid, the time remaining on the auction, and easily list in ranked order the other bids in competition

### PR DESCRIPTION
**Pull request by dckc**
closes: #7266

note the [revised scope](https://github.com/Agoric/agoric-sdk/issues/7266#issuecomment-1489168713)

## Description

Add an `agops inter auction status` command.

### Security/ Scaling Considerations

none?

### Documentation Considerations

BREAKING CHANGE: this replaces an `inter liquidation status` that was kinda goofy.

I'm not sure how `nextDescendingStepTime` works. More on that in a comment below...

### Testing Considerations

The only test is a help text snapshot test.

### CODE REVIEW

**Comment by dckc** [comment]
conflicts already?! I just branched from master a few hours ago. Going to draft mode until I get those worked out...

**Comment by dckc** [comment]
@Chris-Hibbert The [revised scope](https://github.com/Agoric/agoric-sdk/issues/7266#issuecomment-1489168713) scope calls for "Time until next price step". `nextDescendingStepTime` sure seems like it should provide that. I would expect it to change as often as `currentPriceLevel`, but it doesn't. It seems to always agree with `nextStartTime`.

For example:

```
$ date -u && inter auction status
Wed Apr 19 01:30:34 AM UTC 2023
{
  "schedule": {
    "activeStartTime": "2023-04-19T01:30:02.000Z",
    "nextStartTime": "2023-04-19T01:35:02.000Z",
    "nextDescendingStepTime": "2023-04-19T01:30:32.000Z"
  },
  "book0": {
    "startPrice": "9.99 IST/IbcATOM",
    "startProceedsGoal": null,
    "proceedsRaised": null,
    "startCollateral": "0 IbcATOM",
    "collateralAvailable": "0 IbcATOM",
    "currentPriceLevel": "9.4905 IST/IbcATOM"
  },
  "params": {
    "DiscountStep": "5.00%",
    "ClockStep": "00:00:10",
    "LowestRate": "45.00%"
  }
}


$ date -u && inter auction status
Wed Apr 19 01:30:42 AM UTC 2023
{
  "schedule": {
    "activeStartTime": "2023-04-19T01:30:02.000Z",
    "nextStartTime": "2023-04-19T01:35:02.000Z",
    "nextDescendingStepTime": "2023-04-19T01:30:42.000Z"
  },
  "book0": {
    "startPrice": "9.99 IST/IbcATOM",
    "startProceedsGoal": null,
    "proceedsRaised": null,
    "startCollateral": "0 IbcATOM",
    "collateralAvailable": "0 IbcATOM",
    "currentPriceLevel": "8.991 IST/IbcATOM"
  },
  "params": {
    "DiscountStep": "5.00%",
    "ClockStep": "00:00:10",
    "LowestRate": "45.00%"
  }
}
```

cc @rowgraus @otoole-brendan

**Comment by dckc** [comment]
> ... I would expect it to change as often as `currentPriceLevel`, but it doesn't.

oops. yes it does. never mind :)

**Comment by dckc** [comment]
> The only test is a help text snapshot test.

hm... I ended up writing a more extensive test for #7295; maybe I should move it to this PR.

File: packages/agoric-cli/src/commands/inter.js [hunk 1]
```diff
@@ -64,34 +65,38 @@ const makeFormatters = assets => {
...
-    },
-  } = quote;
-  const price = fmt.price({ numerator: amountOut, denominator: amountIn });
-
-  const amounts = objectMap(
-    { liquidatingCollateral, liquidatingDebt },
-    fmt.amount,
-  );
-  return { ...amounts, price };
+  /** @param {bigint} n */
```
Comment (turadg) [inline packages/agoric-cli/src/commands/inter.js]: took me a minute without context to see what this was doing
```suggestion
  /** @param {bigint} bp */
```

File: packages/agoric-cli/src/commands/inter.js [hunk 2]
```diff
@@ -64,34 +65,38 @@ const makeFormatters = assets => {
...
- * Format amounts in vaultManager metrics for JSON output.
- *
- * @param {*} metrics manager0.metrics
- * @param {*} quote manager0.quote
- * @param {*} assets agoricNames.vbankAssets
- */
-const fmtMetrics = (metrics, quote, assets) => {
-  const fmt = makeFormatters(assets);
-  const { liquidatingCollateral, liquidatingDebt } = metrics;
+  /** @param {import('@agoric/time/src/types.js').TimestampRecord} tr */
```
Comment (turadg) [inline packages/agoric-cli/src/commands/inter.js:65]: I think these types are exported at package level
```suggestion
  /** @param {import('@agoric/time').TimestampRecord} tr */
```

File: packages/agoric-cli/src/commands/inter.js [hunk 3]
```diff
@@ -235,34 +240,105 @@ export const makeInterCommand = (
...
+         * @type {{
+         *   book: number,
+         * }}
+         */ opts,
+      ) => {
+        const { agoricNames, readLatestHead } = await tryMakeUtils();
+
+        /**
+         * ref publishSchedule()
+         * packages/inter-protocol/src/auction/scheduler.js#L103
```
Comment (turadg) [inline packages/agoric-cli/src/commands/inter.js]: this line number will be wrong soon enough that it should be omitted. Same for the source path less urgently. For a durable ref try,
```suggestion
         * @see (import('@agoric/inter-protocol/src/auction/scheduler.js')} publishSchedule
```
I don't think our tools error on bad paths there yet but they will

Comment (turadg) [reply packages/agoric-cli/src/commands/inter.js]: oh but more importantly… none of these types should be defined here. The contract must declare its published types and it does.
https://github.com/Agoric/agoric-sdk/blob/470ca4121d46170665958cd713ec9d22dac5eb9a/packages/inter-protocol/src/auction/scheduler.js#L43-L52
https://github.com/Agoric/agoric-sdk/blob/470ca4121d46170665958cd713ec9d22dac5eb9a/packages/inter-protocol/src/auction/auctionBook.js#L93-L107

**Review by turadg — COMMENTED** [review]
> more extensive test for https://github.com/Agoric/agoric-sdk/pull/7295; maybe I should move it to this PR.

I support that since the other PR is about docs. But I'd suggest having the best known tests in both and it'll reach master in whichever lands first.

File: packages/agoric-cli/src/commands/inter.js (hunk 3 above)
Comment (dckc) [reply packages/agoric-cli/src/commands/inter.js]: Thanks! I did just a tiny bit of looking for them, but didn't see them when I hovered over the point of publication.

File: packages/agoric-cli/src/commands/inter.js (hunk 2 above)
Comment (dckc) [reply packages/agoric-cli/src/commands/inter.js:65]: Thanks... I just accepted what I got from ctrl-space

File: packages/agoric-cli/src/commands/inter.js (hunk 3 above)
Comment (turadg) [reply packages/agoric-cli/src/commands/inter.js]: yeah there's some type flow regression in `makeRecorderKit`. I'll look into it.

Comment (turadg) [reply packages/agoric-cli/src/commands/inter.js]: fix in 68d4170d38622089ac3b24f5bf40c1fd0783dff9

**Comment by Chris-Hibbert** [comment]
This all looks good to me. I won't `approve` since @turadg is the listed reviewer.

File: packages/agoric-cli/src/commands/inter.js [hunk 4]
```diff
@@ -64,34 +65,38 @@ const makeFormatters = assets => {
...
-  } = quote;
-  const price = fmt.price({ numerator: amountOut, denominator: amountIn });
-
-  const amounts = objectMap(
-    { liquidatingCollateral, liquidatingDebt },
-    fmt.amount,
-  );
-  return { ...amounts, price };
+  /** @param {bigint} bp */
+  const basisPoints = n => `${(Number(n) / 100).toFixed(2)}%`;
```
Comment (turadg) [inline packages/agoric-cli/src/commands/inter.js]: sorry I omitted the body from the earlier suggestion.
```suggestion
  const basisPoints = bp => `${(Number(bp) / 100).toFixed(2)}%`;
```

File: packages/agoric-cli/src/commands/inter.js [hunk 5]
```diff
@@ -235,34 +240,105 @@ export const makeInterCommand = (
...
+         * @typedef {{
+         *   startPrice: Ratio | null,
+         *   startProceedsGoal: Amount<'nat'> | null,
+         *   proceedsRaised: Amount<'nat'> | null,
+         *   startCollateral: Amount<'nat'>,
+         *   collateralAvailable: Amount<'nat'> | null,
+         *   currentPriceLevel: Ratio | null,
+         * }} BookUpdate
+         */
+        /** @type { [ScheduleUpdate, BookUpdate, *] } */
```
Comment (turadg) [inline packages/agoric-cli/src/commands/inter.js]: I see another push after this discussion but I don't see the types imported. 
```suggestion
         */
        /** @type { [ScheduleNotification, BookDataNotification, *] } */
```

**Review by turadg — CHANGES_REQUESTED** [review]

File: packages/agoric-cli/src/commands/inter.js (hunk 5 above)
Comment (dckc) [reply packages/agoric-cli/src/commands/inter.js]: yeah, I'm still working on it. thanks for watching carefully

Comment (dckc) [reply packages/agoric-cli/src/commands/inter.js]: take another look, please? I hope you don't mind that I squashed all the fixups

Comment (turadg) [reply packages/agoric-cli/src/commands/inter.js]: small enough that it's easy to fully re-review

**Review by turadg — APPROVED** [review]

File: packages/agoric-cli/src/commands/inter.js (hunk 3 above)
Comment (dckc) [reply packages/agoric-cli/src/commands/inter.js]: ok, I applied that in this PR. Didn't manage to preserve authorship, though.

**Comment by dckc** [comment]
darn it... `exit 2 from: make scenario2-run-rosetta-ci'`

File: packages/agoric-cli/src/commands/inter.js (hunk 2 above)
Comment (dckc) [reply packages/agoric-cli/src/commands/inter.js:65]: darn... ci is not happy:

```
Error: src/commands/inter.js(65,38): error TS2694: Namespace '"/home/runner/work/agoric-sdk/agoric-sdk/packages/time/index"' has no exported member 'TimestampRecord'.
```

**Pull request merged**